Block Directory: Uninstall unused block types

ryelle opened the pull request:
## Description
This pulls the rest of #22452 out, which uninstalls any block plugins that were installed from the block directory but not used in this post. See conversation in #22307 for more robust ideas, but this gets the work started with a basic implementation.

The new `AutoBlockUninstaller` component looks through the list of installed blocks (saved in state, `getInstalledBlockTypes`) and filters out any that are not used in the current post. On save, if the post is not published, it loops through this unused block list, and removes those plugins & unregisters the blocks.

The uninstall action is run on every save (except autosaves). The original intent was to be run only on publish, but I think this works better — if you save a draft without using Block A, it gets uninstalled. If you want to use that block after all, you can reinstall it. This approach also avoids leaving plugins in a half-installed state, where abandoned drafts are keeping some plugins installed-but-invisible.

Unfortunately someone can still use the plugin in another post concurrently with this one, otherwise the whole persistence issue would be solved 😆

## How has this been tested?
Install a few blocks from the block directory, and use some in your post. Before saving, open the plugins list in a new tab (or list via wp-cli). All those plugins are installed. Save your post, it should save correctly, no errors. Reload the plugins list, the blocks that weren't used should be removed.

If there was an error uninstalling, it will display a notice in the editor.

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

StevenDufresne committed:
Add unused block uninstall code into its own component.

ryelle committed:
Tweak saving logic

StevenDufresne commented:
What about the PrePublish bar showing the blocks when they are not used in the document? Not sure if that's on purpose. 

![Blocks in sidebar](https://d.pr/i/Ew6EI4.png)

ryelle commented:
> What about the PrePublish bar showing the blocks when they are not used in the document? 

Good catch, I'll update that.

ryelle committed:
Fix mocked API response in test

ryelle committed:
Add new selectors to get used & unused block types

This pulls `hasBlockType` out of the block uninstaller, and moves the logic into selectors so we can re-use it for the "blocks to install" list in pre-publish panel.

ryelle commented:
@StevenDufresne The pre-publish panel should work correctly now. You can even watch blocks get added & removed from the list by opening the panel, and adding/removing installed blocks to your post. Once saved, the blocks not on the list will get uninstalled. I've also added tests for all this logic (in selectors & for `hasBlockType`).

StevenDufresne reviewed (commented):
> @StevenDufresne The pre-publish panel should work correctly now. You can even watch blocks get added & removed from the list by opening the panel, and adding/removing installed blocks to your post. Once saved, the blocks not on the list will get uninstalled. I've also added tests for all this logic (in selectors & for `hasBlockType`).

Yep, it's functioning well. 👍 

I think the uninstall error handling will need another pass.

I also want to bubble up this experience that I reference in #22149 

Notice how the sidebar changes dramatically because the block is no longer installed.

![State flip](https://d.pr/i/fkAfcw.gif)

ryelle commented on packages/block-directory/src/store/test/actions.js:
```diff
@@ -121,4 +121,48 @@ describe( 'actions', () => {
+
+			const apiError = {
+				code: 'could_not_remove_plugin',
+				message: 'Could not fully remove the plugin .',
```
I can change this test, but I wanted to avoid changing the API too much since I think (hope) https://github.com/WordPress/gutenberg/pull/22454 will land soon, and we can iterate again on error flows once it is.

TimothyBJacobs replied:
Yeah if we could minimize or eliminate any changes to the controllers in this PR that'd be awesome :) I'll try my best to make sure that PR lands in the next day or two.

TimothyBJacobs replied:
Yeah if we could minimize or eliminate any changes to the controllers in this PR that'd be awesome :) I'll try my best to make sure that PR lands in the next day or two.

ryelle replied:
@TimothyBJacobs would it help if I removed the `new WP_Error` fixes from this PR?

TimothyBJacobs replied:
That would be!

ryelle replied:
👍 I've rebased (to fix the conflict) and dropped the API changes.

StevenDufresne replied:
👍

ryelle committed:
Update variable name to reflect nested structure

ryelle commented on packages/block-directory/src/components/auto-block-uninstaller/index.js:
```diff
@@ -0,0 +1,71 @@
+
+export default function AutoBlockUninstaller() {
+	const { uninstallBlockType } = useDispatch( 'core/block-directory' );
+	const usedBlocks = useSelect( ( select ) =>
```
`getBlocks` isn't actually returning `blockTypes` ([my var name here is wrong](https://github.com/WordPress/gutenberg/blob/2eb59687a421bb10dd8b8f388d164567873c80ae/packages/block-directory/src/store/selectors.js#L69)), it's returning the representation of blocks in the editor, which includes nested blocks. It's the difference between a `block` — an instance of a block type; and `blockType` — the pattern itself, what we're installing from the block directory.

All that said, I think `usedBlockTree` is more understandable, so I've renamed it  👍

StevenDufresne reviewed (approved):
Looks good! Good job 🏆. Nice to have this back in.
